### PR TITLE
tests: additional coverage for optional config and hyperlink templates

### PR DESCRIPTION
**Pull request by higgsd**
New functionality, new coverage.

# Checklist

If applicable:

- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (`README.md`, `docs/`, `demos/`)
- [ ] I have updated the config schema (`cli/src/config-schema.json`)
- [ ] I have added/updated tests to cover my changes

### CODE REVIEW

File: cli/src/template_builder.rs [hunk 1]
```diff
@@ -2491,14 +2491,21 @@ mod tests {
           |
           = expected `!`, `-`, or <primary>
         ");
-
-        insta::assert_snapshot!(env.parse_err(r#"foo"#), @r"
```
Comment (steadmon) [inline cli/src/template_builder.rs:2571]: What's the reason for removing this one?

Comment (higgsd) [reply cli/src/template_builder.rs:2571]: It seemed redundant with the "!foo" check below, but I don't feel too strongly about it.

Comment (yuja) [reply cli/src/template_builder.rs:2571]: Perhaps, we should instead update the `!foo` test to use `!0` or something? I think it is the test for type error, not invalid symbol.

File: cli/src/template_builder.rs [hunk 2]
```diff
@@ -2491,14 +2491,21 @@ mod tests {
...
+         --> 1:4
           |
-        1 | foo
-          | ^-^
+        1 | 1 +
+          |    ^---
           |
-          = Keyword `foo` doesn't exist
+          = expected `!`, `-`, or <primary>
+        ");
```
Comment (yuja) [inline cli/src/template_builder.rs]: remove duplicated `1 +` test.

File: cli/src/template_builder.rs [hunk 3]
```diff
@@ -3073,6 +3086,10 @@ mod tests {
             env.parse_err_kind("cfg_val == cfg_val"),
             TemplateParseErrorKind::Expression(_)
         );
+        assert_matches!(
+            env.parse_err_kind("some_cfg == some_cfg"),
+            TemplateParseErrorKind::Expression(_)
+        );
```
Comment (yuja) [inline cli/src/template_builder.rs:3169]: (No action requested. Just a comment.)

fwiw, with the current implementation which relies on Rust's type system, I don't think we need extensive tests for negative cases. These tests would be useful when we redesign the template engine in a way that type checking is redone at evaluation stage, though.

**Review by yuja — APPROVED** [review]

File: cli/src/template_builder.rs (hunk 1 above)
Comment (higgsd) [reply cli/src/template_builder.rs:2571]: This kind of type error - including `if(0, true, false)` - is already collected in `test_boolean_cast`.

Should I just remove `!foo` and restore the missing keyword test?

File: cli/src/template_builder.rs (hunk 2 above)
Comment (higgsd) [reply cli/src/template_builder.rs]: Whoops

**Comment by higgsd** [comment]
@yuja can you help clarify a [previous comment](https://github.com/jj-vcs/jj/pull/8725#discussion_r2741787155) of yours?

I went a bit too far and started promoting all syntax and parse errors from `template_builder.rs` unit tests into CLI tests in `test_templater`.  But decided upon further thought that you meant the other way - to keep a "representative" sample of error formatting in `test_templater` and keep the rest as unit tests.

Am I right in this new understanding, and how should I distinguish what error cases are important to keep?  How much  successful template rendering should be kept as a CLI test?  Thanks.

File: cli/src/template_builder.rs (hunk 1 above)
Comment (yuja) [reply cli/src/template_builder.rs:2571]: I don't have strong opinion. I just thought it would be testing type error of unary op because we have things like `true && 123`.

**Comment by yuja** [comment]
> @yuja can you help clarify a [previous comment](https://github.com/jj-vcs/jj/pull/8725#discussion_r2741787155) of yours?
> 
> I went a bit too far and started promoting all syntax and parse errors from `template_builder.rs` unit tests into CLI tests in `test_templater`. But decided upon further thought that you meant the other way - to keep a "representative" sample of error formatting in `test_templater` and keep the rest as unit tests.

Yeah, something like the latter. I meant we need "some" e2e tests that cover the actual error formatting (with hints). That's more important for alias substitution errors because source errors should be printed in a reasonable way.

File: cli/src/template_builder.rs (hunk 1 above)
Comment (higgsd) [reply cli/src/template_builder.rs:2571]: I dropped this hunk entirely and just kept the new `expected <function>` syntax case.  I'll think about what to do with this grab bag of parse errors when I try to address `test_templater`.

**Comment by higgsd** [comment]
Last second change so the description reflects the commit content, since I didn't add any tests for hyperlinks.
